### ISSUE TITLE
WiFi clients may 'stick' around

### ISSUE THREAD

**Issue by batesta** [batesta/whoshere#2]
It's possible, that once a WiFi user leaves, you'll see them on the LAN (wired) network for a length of time.

I tried to code it that if that happens, it doesn't show them there, but the time would be incorrect.

This seems to happen when IPS is turned on for me.

**Comment by spmfox** [comment]
(Disclaimer: I have not looked at the code, and I have not tried the software yet.)

I don't think it's IPS, I see that all the time when wireless clients join and disconnect - sometimes they appear on the LAN for a bit. I've never seen it stay like that for long, but yeah 2 or 5 minutes is common.

My suggestion would be to make the default behavior say any person connected via LAN is actually not connected.

I've not looked at the code, I'm not sure if you're calculating the timestamp from the last time the script saw the client on the controller or the last time the controller saw the client.

If you're basing the time from the controller, there really is no accurate way to avoid the lagging client disconnect. If you're calculating your own timestamp then simply ignore LAN clients by default and perhaps expose an option to not ignore.

I don't know much PHP but I do know weird Ubiquiti issues lol

**Comment by batesta** [comment]
> (Disclaimer: I have not looked at the code, and I have not tried the software yet.)
> 
> I don't think it's IPS, I see that all the time when wireless clients join and disconnect - sometimes they appear on the LAN for a bit. I've never seen it stay like that for long, but yeah 2 or 5 minutes is common.
> 

Yeah, that was kind of an assumption I had about that.  I seemed to only see the issue when IPS was turned on, but it might just be because our Unifi USG4P is too taxed.  

I've tried to 'fix' it, in that if a MAC address is found in the active table, but doesn't have a corresponding hotspot MAC attached to it, they're really not there. BUT.....


> My suggestion would be to make the default behavior say any person connected via LAN is actually not connected.
> 
> I've not looked at the code, I'm not sure if your calculating the timestamp from the last time the script saw the client on the controller or the last time the controller saw the client.
> 
> If your basing the time from the controller, there really is no accurate way to avoid the lagging client disconnect. If your calculating your own timestamp then simply ignore LAN clients by default and perhaps expose an option to not ignore.
> 


..the time is then off, unfortunately. :(  Since I'm just using the UniFi timestamps.


> I don't know much PHP but I do know weird Ubiquiti issues lol

Thanks for looking at it!

**Comment by spmfox** [comment]
Since you're already manually defining access points and clients, why not modify the script to leave a timestamp for each MAC address in a file as well. Put a simple locking mechanism so you don't write it from multiple sessions. If the client shows on LAN then it's considered offline, as well as actually being offline. Then you use your timestamp and ignore the controller's.

Then have the same script run via cron, maybe every five minutes. That will keep your data up-to-date.

It does increase the scope a little, but if the actual timestamp is important then it's an option.

My personal opinion is: if you were looking at the controller history for this client, it would also give you the bad timestamp under the "last seen" time. I'm sure it also counts this "LAN" connection as "seen". If Unifi is giving you bad data, it's not your software's fault. You're doing your job.

I believe if you personally need the timestamps to be right, you should explore my suggestion. If not, then your software is working as expected - pulling the Unifi data and presenting it the way you want.

**Comment by geoffsmith82** [comment]
Pinging the device may be another way of seeing if the device is still on the network

**Comment by spmfox** [comment]
@geoffsmith82 yeah however the type of devices (phones?) being queried may block pings. That's just a hunch but I've never checked it on a large scale.

**Comment by geoffsmith82** [comment]
I did a quick check with my phone and it seemed to work.  Of course this doesn't mean to say it will work on every phone... and you would need to make sure the IP and the MAC address still match - a new device could have taken over the IP - although that would just result in the existing 'sticky' behavior.

**Comment by abbeydental** [comment]
I tried running your code on my machine but I keep getting...

 Notice: Trying to get property 'last_seen' of non-object in whoshere.php on line 68

Any idea what's going on ?


------ Original Message ------

**Comment by batesta** [comment]
Are you sure the script can login to your UnFi controller successfully?

It also could be the MAC address you specified doesn't exist on the UniFi controller.
